doc/sg_rep_density.8
.TH SG_REP_DENSITY "8" "June 2023" "sg3_utils\-1.48" SG3_UTILS
.SH NAME
sg_rep_density \- send SCSI REPORT DENSITY SUPPORT command
.SH SYNOPSIS
.B sg_rep_density
[\fI\-\-help\fR] [\fI\-\-hex\fR] [\fI\-\-inhex=FN\fR] [\fI\-\-maxlen=LEN\fR]
[\fI\-\-media\fR] [\fI\-\-raw\fR] [\fI\-\-readonly\fR] [\fI\-\-typem\fR]
[\fI\-\-verbose\fR] [\fI\-\-version\fR]
\fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
Sends a SCSI REPORT DENSITY command to \fIDEVICE\fR and outputs the data
returned. This command is tape drive specific. This command is found in
the SSC\-5 draft standard, revision 6 (ssc5r06.pdf). This command was
present in the SSC\-2 standard (ANSI INCITS 380\-2003).
.PP
By default this utility requests the density code descriptors supported by
the \fIDEVICE\fR (e.g. a tape drive) and decodes the response. If the
\fI\-\-typem\fR option is given it fetches the medium type descriptors
supported by the \fIDEVICE\fR and decodes the response. When the
\fI\-\-media\fR option is given the density code or medium type descriptors
supported by the media inside the \fIDEVICE\fR (e.g. a tape cartridge) are
fetched.
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-H\fR, \fB\-\-hex\fR
output the response in hexadecimal to stdout. When used once the whole
response is output in ASCII hexadecimal, prefixed by an address (starting at
0) on each line. When used twice the whole response is output in hexadecimal
with no leading address (on each line).
.br
Using this option three times will produce output that can be redirected to
a file and later given to another invocation using the \fI\-\-inhex=FN\fR
option.
.TP
\fB\-i\fR, \fB\-\-inhex\fR=\fIFN\fR
where \fIFN\fR is a file name whose contents are assumed to be ASCII
hexadecimal. If \fIDEVICE\fR is also given then \fIDEVICE\fR is ignored,
a warning is issued and the utility continues, decoding the file named
\fIFN\fR. See the "HEX, BINARY AND JSON FORMATS" section in the
sg3_utils manpage for more information. If the \fI\-\-raw\fR option is
also given then the contents of \fIFN\fR are treated as binary.
.TP
\fB\-m\fR, \fB\-\-maxlen\fR=\fILEN\fR
where \fILEN\fR is the (maximum) response length in bytes. It is placed in
the cdb's "allocation length" field. If not given (or \fILEN\fR is zero)
then 8192 is used. The maximum allowed value of \fILEN\fR is 65535.
.TP
\fB\-M\fR, \fB\-\-media\fR
sets the MEDIA bit in the cdb which causes the density codes (or medium
types) supported by the tape cartridge in the drive to be placed in the
response. The default is to request the density codes (or medium types)
supported by the tape drive itself.
.br
If there is no "medium" (e.g. tape cartridge) present in the drive the SCSI
command will fail with a "not ready" sense key.
.TP
\fB\-r\fR, \fB\-\-raw\fR
output the SCSI response (i.e. the data\-out buffer) in binary (to stdout)
unless the \fI\-\-inhex=FN\fR option is given.
.br
When used together with the \fI\-\-inhex=FN\fR option then the contents of
\fIFN\fR are treated as binary (rather than hexadecimal).
.TP
\fB\-R\fR, \fB\-\-readonly\fR
open the \fIDEVICE\fR read\-only (e.g. in Unix with the O_RDONLY flag).
The default is to open it read\-write.
.TP
\fB\-t\fR, \fB\-\-typem\fR
sets the MEDIUM TYPE bit in the cdb which causes the medium types supported
by the tape drive (or tape cartridge) to be placed in the response. The
default is to request the density codes.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH EXIT STATUS
The exit status of sg_rep_density is 0 when it is successful. Otherwise see
the sg3_utils(8) man page.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2022\-2023 Douglas Gilbert
.br
This software is distributed under a BSD\-2\-Clause license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sg3_utils(sg3_utils)


CHANGELOG
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for sg3_utils-1.25 [20070710]
  - sg_dd: add oflag=sparse to step over bs*bpt number of zeros;
    - with oflag=sparse, write last bs*bpt segment at end or after
      error so file length of OFILE is appropriate
    - allow iflag=null and oflag=null which do nothing (placeholders)
    - when coe>1 then SCSI READ LONG logic remember extended block
      length of first encountered error
  - sg_ses: sync with ses2r17
  - place source in subversion repository

Changelog for sg3_utils-1.24 [20070507]
  - sg_raw: new utility to send arbirary SCSI commands
  - sg_luns: increase number of luns that can be fetched
    - fix length of raw and hex output
    - add '--quiet' option to output only ASCII hex
      representation of each lun
  - sg_rtpg: update for changes in spc4r09
  - sg_persist: update documentation, spc-4 references
    - fix exit status values
  - sg_inq: update version descriptors per spc4r09
    - fix '--id' and '--extended'
    - extend block limits VPD page (sbc3r09)
  - sg_vpd: extend block limits VPD page (sbc3r09)
    - append relative target port identifier to SAS target
      port address with '-iq' option
  - sg_logs: add decoding for stats+performance log pages
    - fix showing of page names for pdt > 0
    - implement '-HH' for single and all pages, fix '-r'
    - when '--maxlen=' given, only do single fetch
    - add Tape Alert (ssc), Media and Element statistics (smc) pages 
    - add '--brief' option
  - sg_ses: sync with ses2r16
    - fix bay number for SAS
  - sg_format: add '--dcrt' and '--security' options
  - sgm_dd: add 'smmap' oflag for shared_mmap_io testing
    - add 'dio' oflag
  - sg_dd, sgp_dd: add 'dio' iflag and oflag
  - sg_modes: change SAS mode page names per sas2r09
    - check validity of block descriptors length
  - sg_pt: change opaque context object from 'void *'
    to 'struct sg_pt_base *'
  - sg_opcodes: anticipate extra tmfs from 07-159r0
  - sg_sat_set_features: add more usage information
    - add man page
  - sg_sat_phy_event: add to examples directory
  - sg_lib: sync asc/ascq codes with spc4r10
  - Solaris port: using uscsi interface

Changelog for sg3_utils-1.23 [20070131]
  - sg_read_buffer: new utility
  - sg_write_buffer: new utility
  - sg_opcodes, sg_senddiag, sg_logs, sg_modes, sg_start, sg_inq,
    sg_turs, sg_readcap, sg_rbuf: add getopt_long() based cli;
    old and new cli selectable, new getopt_long cli is default
  - scripts: new subdirectory containing some bash scripts
    - add scripts/README file
  - sg_reassign: add '--hex' option for grown and primary lists
  - sg_rtpg: add '--raw' option
  - sg_lib.h, sg_cmds_basic.h + sg_cmds_extra.h: add C++ 
    'extern "C" ' wrappers
    - cleanup C code so it will compile as C++
  - sg_lib: sync with spc4r08
    - include <inttypes.h>, use PRId64 instead of %lld form
    - fix sg_get_sense_str() when empty sense buffer
  - win32 port: add Makefile.mingw + related support for MinGW
  - sg_cmds_extra: add sg_ll_read_buffer() and sg_ll_write_buffer()
  - sg_dd, sgp_dd, sgm_dd, sg_read: use lseek64() instead of llseek.c
  - sgm_dd: accept coe=<n> for interworking with sg_dd
  - sg_rdac: fix on non-linux ports
  - sg_ses: fix spurious warning in additional element status page
    - '-rr' option outputs a diagnostic page in binary to stdout
  - sg_opcodes: add command timeout descriptor support (spc4r08)
    - change linux specific pass through to generic pass through
  - sg_logs: add 'name=value' decoding for SAS specific lpage
  - examples+utils subdirectories: remove symlinks
  - synchronize man pages with usage messages
  - sg3_utils.spec: rework

Changelog for sg3_utils-1.22 [20061016]
  - sgp_dd: accept verbose=<n> as well as deb=<n> to ease
    interworking with sg_dd and sgm_dd
  - sg_sat_set_features: added to examples directory
  - sg_lib: sync asc/ascq text with spc4r06
    - move SG_LIB_CAT_NO_SENSE and SG_LIB_CAT_RECOVERED to
      20 and 21 respectively; add SG_LIB_CAT_ABORTED_COMMAND
      at 11 (its sense key value) 
  - sg_vpd: tweak '--page=sp --quiet' output
    - change '-HHH' so same as '-rr' (prepares ATA Information
      (ai) response for hdparm)
  - sg_requests: add '-s' option to set exit status from
    parameter data
  - sg_modes: exit quickly from '-e' if device not ready
  - sg_logs: sync sas log pages with sas2r05a
    - expand background scan results log page
    - add '-m=<max_len>' to limit response length
    - drop '-scum' and '-sthr' options and add '-select'
  - sg_write_long: add '--16' option to send 16 byte cdb
    - add '--wr_uncor' and '--pblock' options
  - sg_senddiag: cleanup and add sdiag_sas_p1_stop.txt
    to examples directory
  - sg_format: add '--cmplst=<n>' option (default: 1)
    - add '--pfu=<n>' option
    - expand man page to discuss P/D/C/GLISTs
  - sg_reassign: add '--primary' option to fetch primary
    defect list (PLIST) length
  - sg_readcap: add '-H' option to output response in hex
    and '-r' to output response in binary to stdout
    - add logical blocks per physical block (sbc3r07)
  - sginfo: add PLIST and GLIST designation to defect lists
  - sg_cmds: split this support file into sg_cmds_basic.[hc]
    and sg_cmds_extra.[hc]
    - add sg_ll_ata_pt() (SATL ATA pass) to sg_cmds_extra.[hc]
  - sg_rdac: fix includes for FreeBSD
  - sg_dd: add 'coe_limit=' option to exit after <n>
    consecutive 'coe' type read errors
  - sgm_dd: print out throughput information when signal arrives
    if time=1 (like sg_dd does already)
  - sg_inq: change '-HHH' so same as '-rr'. Now sg_inq, sg_vpd
    and sdparm output for hdparm with '-HHH'
    -add '-l=<resp_len>' option
  - sg_read_long: add '--pblock' option for physical blocks
  - sg_luns: add '--hex' and '--raw' options

src/sg_rep_density.c
/*
 * Copyright (c) 2022-2023 Douglas Gilbert.
 * All rights reserved.
 * Use of this source code is governed by a BSD-style
 * license that can be found in the BSD_LICENSE file.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <getopt.h>
#define __STDC_FORMAT_MACROS 1
#include <inttypes.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "sg_lib.h"
#include "sg_lib_data.h"
#include "sg_pt.h"
#include "sg_cmds_basic.h"
#include "sg_unaligned.h"
#include "sg_pr2serr.h"

/* A utility program originally written for the Linux OS SCSI subsystem.
 *
 *
 * This program issues the SCSI REPORT DENSITY SUPPORT command to the given
 * SCSI (tape) device and outputs the response. Based on ssc5r06.pdf
 */

static const char * version_str = "1.02 20230519";

static const char * my_name = "sg_rep_density: ";

#define MAX_RDS_BUFF_LEN (64 * 1024 - 1)
#define DEF_RDS_BUFF_LEN 4096

#define REPORT_DENSITY_SUPPORT_CMD 0x44
#define REPORT_DENSITY_SUPPORT_CMDLEN 10

#define RDS_DENSITY_DESC_LEN 52
#define RDS_MEDIUM_T_DESC_LEN 56

#define SENSE_BUFF_LEN 64       /* Arbitrary, could be larger */
#define DEF_PT_TIMEOUT  60      /* 60 seconds */

static const char * rds_s = "Report density support";


static struct option long_options[] = {
        {"help", no_argument, 0, 'h'},
        {"hex", no_argument, 0, 'H'},
        {"in", required_argument, 0, 'i'},      /* silent, same as --inhex= */
        {"inhex", required_argument, 0, 'i'},
        {"maxlen", required_argument, 0, 'm'},
        {"media", no_argument, 0, 'M'}, /* Media field; byte 1, bit 0 */
        {"raw", no_argument, 0, 'r'},
        {"readonly", no_argument, 0, 'R'},
        {"typem", no_argument, 0, 't'}, /* Medium type field, byte 1, bit 1 */
        {"verbose", no_argument, 0, 'v'},
        {"version", no_argument, 0, 'V'},
        {0, 0, 0, 0},
};


static void
usage(void)
{
    pr2serr("Usage: "
            "sg_rep_density  [--help] [--hex] [--inhex=FN] [--maxlen=LEN] "
            "[--media]\n"
            "                   [--raw] [--readonly] [--typem] [--verbose] "
            "[--version]\n"
            "                   DEVICE\n");
    pr2serr("  where:\n"
            "    --help|-h          prints out this usage message\n"
            "    --hex|-H           output response in hexadecimal "
            "(default); used\n"
            "                       twice: hex without addresses at start "
            "of line\n"
            "    --inhex=FN         decode contents of FN, ignore DEVICE\n"
            "    --maxlen=LEN|-m LEN    max response length (allocation "
            "length in cdb)\n"
            "                           (def: 512 bytes)\n"
            "    --media|-M         report on media in drive (def: report "
            "on drive)\n"
            "    --raw|-r           output response in binary\n"
            "    --readonly|-R      open DEVICE read-only (def: read-write)\n"
            "    --typem|-t         report medium types (def: density "
            "codes)\n"
            "    --verbose|-v       increase verbosity\n"
            "    --version|-V       print version string and exit\n\n"
            "Sends a SCSI REPORT DENSITY SUPPORT command outputs the "
            "response in\nASCII hexadecimal or binary. By default it reports "
            "on density codes supported\nby the drive (LU).\n");
}

/* Invokes a SCSI REPORT PROVISIONING INITIALIZATION PATTERN command (SBC).
 * Return of 0 -> success, various SG_LIB_CAT_* positive values or
 * -1 -> other errors */
static int
sg_ll_report_density(int sg_fd, bool media, bool m_type, void * resp,
                     int mx_resp_len, int * residp, bool noisy, int verbose)
{
    int ret, res, sense_cat;
    uint8_t rds_cdb[REPORT_DENSITY_SUPPORT_CMDLEN] =
          {REPORT_DENSITY_SUPPORT_CMD, 0, 0, 0,  0, 0, 0, 0,  0, 0};
    uint8_t sense_b[SENSE_BUFF_LEN] SG_C_CPP_ZERO_INIT;
    struct sg_pt_base * ptvp;

    if (media)
        rds_cdb[1] |= 0x1;
    if (m_type)
        rds_cdb[1] |= 0x2;
    sg_put_unaligned_be16((uint16_t)mx_resp_len, rds_cdb + 7);
    if (verbose) {
        char b[128];

        pr2serr("    %s cdb: %s\n", rds_s,
                sg_get_command_str(rds_cdb, REPORT_DENSITY_SUPPORT_CMDLEN,
                                   false, sizeof(b), b));
    }
    ptvp = construct_scsi_pt_obj();
    if (NULL == ptvp) {
        pr2serr("%s: out of memory\n", __func__);
        return -1;
    }
    set_scsi_pt_cdb(ptvp, rds_cdb, sizeof(rds_cdb));
    set_scsi_pt_sense(ptvp, sense_b, sizeof(sense_b));
    set_scsi_pt_data_in(ptvp, (uint8_t *)resp, mx_resp_len);
    res = do_scsi_pt(ptvp, sg_fd, DEF_PT_TIMEOUT, verbose);
    ret = sg_cmds_process_resp(ptvp, rds_s, res, noisy, verbose, &sense_cat);
    if (-1 == ret) {
        if (get_scsi_pt_transport_err(ptvp))
            ret = SG_LIB_TRANSPORT_ERROR;
        else
            ret = sg_convert_errno(get_scsi_pt_os_err(ptvp));
    } else if (-2 == ret) {
        switch (sense_cat) {
        case SG_LIB_CAT_RECOVERED:
        case SG_LIB_CAT_NO_SENSE:
            ret = 0;
            break;
        default:
            ret = sense_cat;
            break;
        }
    } else
        ret = 0;
    if (residp)
        *residp = get_scsi_pt_resid(ptvp);
    destruct_scsi_pt_obj(ptvp);
    return ret;
}

static void
decode_medium_type(const uint8_t * up, int num_desc)
{
    int k, j, n, q;

    for (k = 0; k < num_desc; ++k, up += RDS_MEDIUM_T_DESC_LEN) {
        if (0 == k)
            printf("Medium type descriptor%s\n", ((num_desc > 1) ? "s" : ""));
        printf("  descriptor %d\n", k + 1);
        printf("    Medium type: %u\n", up[0]);
        n = up[4];
        printf("    Number of density codes: %d\n", n);
        if (n > 9)
            n = 9;
        for (j = 0; j < n; ++j) {
            q = up[5 + j];
            if (q > 0)
                printf("      Primary density code: %d\n", q);
        }
        printf("    Media width: %u\n", sg_get_unaligned_be16(up + 14));
        printf("    Medium length: %u\n", sg_get_unaligned_be16(up + 16));
        printf("    Assigning organization: %.8s\n", (const char *)(up + 20));
        printf("    Medium type name: %.8s\n", (const char *)(up + 28));
        printf("    Description: %.20s\n", (const char *)(up + 36));
    }
}

static void
decode_density_code(const uint8_t * up, int num_desc)
{
    int k;

    for (k = 0; k < num_desc; ++k, up += RDS_DENSITY_DESC_LEN) {
        if (0 == k)
            printf("Density support data block descriptor%s\n",
                   ((num_desc > 1) ? "s" : ""));
        printf("  descriptor %d\n", k + 1);
        printf("    Primary density code: %u\n", up[0]);
        printf("    Secondary density code: %u\n", up[1]);
        printf("    WRT: %u\n", !!(0x80 & up[2]));
        printf("    DUP: %u\n", !!(0x40 & up[2]));
        printf("    DEFLT: %u\n", !!(0x20 & up[2]));
        printf("    DLV: %u\n", !!(0x1 & up[2]));
        printf("    Bits per mm: %u\n", sg_get_unaligned_be24(up + 5));
        printf("    Media width: %u\n", sg_get_unaligned_be16(up + 8));
        printf("    Tracks: %u\n", sg_get_unaligned_be16(up + 10));
        printf("    Capacity: %u\n", sg_get_unaligned_be32(up + 12));
        printf("    Assigning organization: %.8s\n", (const char *)(up + 16));
        printf("    Density name: %.8s\n", (const char *)(up + 24));
        printf("    Description: %.20s\n", (const char *)(up + 32));
    }
}

static void
dStrRaw(const uint8_t * str, int len)
{
    int k;

    for (k = 0; k < len; ++k)
        printf("%c", str[k]);
}


int
main(int argc, char * argv[])
{
    bool do_raw = false;
    bool media = false;
    bool m_type = false;
    bool no_final_msg = false;
    bool o_readonly = false;
    bool verbose_given = false;
    bool version_given = false;
    int res, c, rlen, desc_len, ads_len, num_desc;
    int resid = 0;
    int sg_fd = -1;
    int do_help = 0;
    int do_hex = 0;
    int maxlen = 0;
    int in_len = 0;
    int ret = 0;
    int verbose = 0;
    const char * device_name = NULL;
    const char * inhex_fn = NULL;
    uint8_t * rdsBuff = NULL;
    uint8_t * free_rds = NULL;
    char b[80];

    if (getenv("SG3_UTILS_INVOCATION"))
        sg_rep_invocation(my_name, version_str, argc, argv, stderr);
    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "hHi:m:MrRtvV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'h':
        case '?':
            ++do_help;
            break;
        case 'H':
            ++do_hex;
            break;
        case 'i':
            inhex_fn = optarg;
            break;
        case 'm':
            maxlen = sg_get_num(optarg);
            if ((maxlen < 0) || (maxlen > MAX_RDS_BUFF_LEN)) {
                pr2serr("argument to '--maxlen' should be %d or less\n",
                        MAX_RDS_BUFF_LEN);
                return SG_LIB_SYNTAX_ERROR;
            }
            break;
        case 'M':
            media = true;
            break;
        case 'r':
            do_raw = true;
            break;
        case 'R':
            o_readonly = true;
            break;
        case 't':
            m_type = true;
            break;
        case 'v':
            verbose_given = true;
            ++verbose;
            break;
        case 'V':
            version_given = true;
            break;
        default:
            pr2serr("unrecognised option code 0x%x ??\n", c);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if (NULL == device_name) {
            device_name = argv[optind];
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                pr2serr("Unexpected extra argument: %s\n", argv[optind]);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
#ifdef DEBUG
    pr2serr("In DEBUG mode, ");
    if (verbose_given && version_given) {
        pr2serr("but override: '-vV' given, zero verbose and continue\n");
        /* verbose_given = false; */
        version_given = false;
        verbose = 0;
    } else if (! verbose_given) {
        pr2serr("set '-vv'\n");
        verbose = 2;
    } else
        pr2serr("keep verbose=%d\n", verbose);
#else
    if (verbose_given && version_given)
        pr2serr("Not in DEBUG mode, so '-vV' has no special action\n");
#endif
    if (version_given) {
        pr2serr("version: %s\n", version_str);
        return 0;
    }

    if (do_help) {
        usage();
        return 0;
    }
    if (device_name && inhex_fn) {
        pr2serr("ignoring DEVICE, best to give DEVICE or --inhex=FN, but "
                "not both\n");
        device_name = NULL;
    }
    if (0 == maxlen)
        maxlen = DEF_RDS_BUFF_LEN;
    rdsBuff = (uint8_t *)sg_memalign(maxlen, 0, &free_rds, verbose > 3);
    if (NULL == rdsBuff) {
        pr2serr("unable to sg_memalign %d bytes\n", maxlen);
        return sg_convert_errno(ENOMEM);
    }
    if (NULL == device_name) {
        if (inhex_fn) {
            if ((ret = sg_f2hex_arr(inhex_fn, do_raw, false, rdsBuff,
                                    &in_len, maxlen))) {
                if (SG_LIB_LBA_OUT_OF_RANGE == ret) {
                    no_final_msg = true;
                    pr2serr("... decode what we have, --maxlen=%d needs to "
                            "be increased\n", maxlen);
                } else
                    goto the_end;
            }
            if (verbose > 2)
                pr2serr("Read %d [0x%x] bytes of user supplied data\n",
                        in_len, in_len);
            if (do_raw)
                do_raw = false;    /* otherwise interferes with decoding */
            if (in_len < 4) {
                pr2serr("--inhex=%s only decoded %d bytes (needs 4 at "
                        "least)\n", inhex_fn, in_len);
                ret = SG_LIB_SYNTAX_ERROR;
                goto the_end;
            }
            res = 0;
            maxlen = in_len;
            goto start_response;
        } else {
            pr2serr("missing device name!\n\n");
            usage();
            ret = SG_LIB_FILE_ERROR;
            no_final_msg = true;
            goto the_end;
        }
    } else
        in_len = 0;

    if (do_raw) {
        if (sg_set_binary_mode(STDOUT_FILENO) < 0) {
            perror("sg_set_binary_mode");
            ret = SG_LIB_FILE_ERROR;
            no_final_msg = true;
            goto the_end;
        }
    }

    sg_fd = sg_cmds_open_device(device_name, o_readonly, verbose);
    if (sg_fd < 0) {
        if (verbose)
            pr2serr("open error: %s: %s\n", device_name,
                    safe_strerror(-sg_fd));
        ret = sg_convert_errno(-sg_fd);
        goto the_end;
    }

    res = sg_ll_report_density(sg_fd, media, m_type, rdsBuff, maxlen, &resid,
                               true, verbose);
start_response:
    ret = res;
    if (0 == res) {
        rlen = maxlen - resid;
        if (rlen < 4) {
            pr2serr("Response length (%d) too short\n", rlen);
            ret = SG_LIB_CAT_MALFORMED;
            goto the_end;
        }
        if (do_raw) {
            dStrRaw(rdsBuff, rlen);
            goto the_end;
        }
        if (do_hex) {
            if (2 != do_hex)
                hex2stdout(rdsBuff, rlen, ((1 == do_hex) ? 1 : -1));
            else
                hex2stdout(rdsBuff, rlen, 0);
            goto the_end;
        }
        desc_len = m_type ? RDS_MEDIUM_T_DESC_LEN : RDS_DENSITY_DESC_LEN;
        ads_len = sg_get_unaligned_be16(rdsBuff + 0) + 2;
        if (4 == ads_len)
            goto the_end;
        if (ads_len < 4) {
            pr2serr("Badly formatted response, ads_len=%d\n", ads_len - 2);
            ret = SG_LIB_CAT_MALFORMED;
            goto the_end;
        }
        if (ads_len > rlen) {
            if (verbose)
                pr2serr("Trimming response from %d to %d bytes\n", ads_len,
                        rlen);
            ads_len = rlen;
            if (4 == ads_len)
                goto the_end;
        }
        num_desc = (ads_len - 4) / desc_len;
        if (0 != ((ads_len - 4) % desc_len)) {
            if (verbose)
                pr2serr("Truncating response to %d descriptors\n", num_desc);
        }
        if (m_type)
            decode_medium_type(rdsBuff + 4, num_desc);
        else
            decode_density_code(rdsBuff + 4, num_desc);
    } else if (SG_LIB_CAT_INVALID_OP == res)
        pr2serr("%s command not supported\n", rds_s);
    else {
        sg_get_category_sense_str(res, sizeof(b), b, verbose);
        pr2serr("%s command: %s\n", rds_s, b);
    }

the_end:
    if (free_rds)
        free(free_rds);
    if (sg_fd >= 0) {
        res = sg_cmds_close_device(sg_fd);
        if (res < 0) {
            pr2serr("close error: %s\n", safe_strerror(-res));
            if (0 == ret)
                ret = sg_convert_errno(-res);
        }
    }
    if ((0 == verbose) && (! no_final_msg)) {
        if (! sg_if_can2stderr("sg_rep_density failed: ", ret))
            pr2serr("Some error occurred, try again with '-v' "
                    "or '-vv' for more information\n");
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
